authorize/evaluator/headers_evaluator_evaluation.go

```diff
@@ -0,0 +1,476 @@
+			"user":   e.getJWTPayloadUser(ctx),
+			"email":  e.getJWTPayloadEmail(ctx),
+			"groups": e.getJWTPayloadGroups(ctx),
+			"sid":    e.getJWTPayloadSID(),
+			"name":   e.getJWTPayloadName(ctx),
+		}
+
+		s, _ := e.getSessionOrServiceAccount(ctx)
+		u := e.getUser(ctx)
+
+		for _, claimKey := range e.evaluator.store.GetJWTClaimHeaders() {
+			// ignore base claims
+			if _, ok := e.cachedJWTPayload[claimKey]; ok {
+				continue
+			}
+
+			if vs, ok := getClaimStringSlice(s, claimKey); ok {
+				e.cachedJWTPayload[claimKey] = strings.Join(vs, ",")
+			} else if vs, ok := getClaimStringSlice(u, claimKey); ok {
+				e.cachedJWTPayload[claimKey] = strings.Join(vs, ",")
+			}
+		}
+	}
+	return e.cachedJWTPayload
+}
+
+func (e *headersEvaluatorEvaluation) getSignedJWT(ctx context.Context) string {
+	if !e.gotSignedJWT {
+		e.gotSignedJWT = true
+		signingKey := e.evaluator.store.GetSigningKey()
+		if signingKey == nil {
+			log.Ctx(ctx).Error().Msg("authorize/header-evaluator: missing signing key")
+			return ""
+		}
+
+		signingOptions := new(jose.SignerOptions).
+			WithType("JWT").
+			WithHeader("kid", signingKey.KeyID).
+			WithHeader("alg", signingKey.Algorithm)
+
+		signer, err := jose.NewSigner(jose.SigningKey{
+			Algorithm: jose.SignatureAlgorithm(signingKey.Algorithm),
+			Key:       signingKey.Key,
+		}, signingOptions)
+		if err != nil {
+			log.Ctx(ctx).Error().Err(err).Msg("authorize/header-evaluator: error creating JWT signer")
+			return ""
+		}
+
+		jwtPayload := e.getJWTPayload(ctx)
+		bs, err := json.Marshal(jwtPayload)
+		if err != nil {
+			log.Ctx(ctx).Error().Err(err).Msg("authorize/header-evaluator: error marshaling JWT payload")
+			return ""
+		}
+
+		jwt, err := signer.Sign(bs)
+		if err != nil {
+			log.Ctx(ctx).Error().Err(err).Msg("authorize/header-evaluator: error signing JWT")
+			return ""
+		}
+
+		e.cachedSignedJWT, err = jwt.CompactSerialize()
+		if err != nil {
+			log.Ctx(ctx).Error().Err(err).Msg("authorize/header-evaluator: error serializing JWT")
+			return ""
+		}
+	}
+
+	return e.cachedSignedJWT
+}
+
+func (e *headersEvaluatorEvaluation) getDataBrokerGroupNames(ctx context.Context, groupIDs []string) []string {
+	groupNames := make([]string, 0, len(groupIDs))
+	for _, groupID := range groupIDs {
+		dg, _ := e.evaluator.store.GetDataBrokerRecord(ctx, directory.GroupRecordType, groupID).(*structpb.Struct)
+		if name, ok := getStructString(dg, "name"); ok {
+			groupNames = append(groupNames, name)
+		}
+	}
+	return groupNames
+}
+
+type hasGetClaims interface {
+	GetClaims() map[string]*structpb.ListValue
+}
+
+func getClaimStringSlice(msg hasGetClaims, field string) ([]string, bool) {
+	if msg == nil {
+		return nil, false
+	}
+
+	claims := msg.GetClaims()
+	if claims == nil {
+		return nil, false
+	}
+
+	lv, ok := claims[field]
+	if !ok {
+		return nil, false
+	}
+
+	strs := make([]string, len(lv.Values))
+	for i, v := range lv.Values {
+		strs[i] = fmt.Sprint(v.AsInterface())
```

Comment:
Do we want to support object/list claim values? This will serialize objects into a format that looks like `map[key:value key2:value2]`. The existing behavior is that if you have any object or list in the claim value, it skips the entire claim, but that doesn't seem correct either.

Reply:
I updated the code to just skip them.